joeynmt/transformer.py

```diff
@@ -0,0 +1,233 @@
+        Compute multi-headed attention.
+
+        :param query:
+        :param key:
+        :param value:
+        :param mask:
+        :return: new x
+        """
+
+        if mask is not None:
+            # Same mask applied to all h heads.
+            mask = mask.unsqueeze(1)
+        nbatches = query.size(0)
+
+        # 1) Do all the linear projections in batch from d_model => h x d_k
+        query, key, value = \
+            [l(x).view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+             for l, x in zip(self.linears, (query, key, value))]
+
+        # 2) Apply attention on all the projected vectors in batch.
+        x, self.attn = attention(query, key, value, mask=mask,
+                                 dropout=self.dropout)
+
+        # 3) "Concat" using a view and apply a final linear.
+        x = x.transpose(1, 2).contiguous() \
+            .view(nbatches, -1, self.h * self.d_k)
+        return self.linears[-1](x)
+
+
+class PositionwiseFeedForward(nn.Module):
+    """
+    Position-wise Feed-forward layer
+    Projects to d_ff and then back down to d_model.
+    """
+
+    def __init__(self, d_model, d_ff, dropout=0.1):
+        """
+        Initializes position-wise feed-forward layer.
+        :param d_model: dimensionality of the input.
+        :param d_ff: dimensionality of intermediate representation
+        :param dropout:
+        """
+        super(PositionwiseFeedForward, self).__init__()
+        self.layer = nn.Sequential(
+            nn.Linear(d_model, d_ff),
+            nn.ReLU(),
+            nn.Dropout(dropout),
+            nn.Linear(d_ff, d_model)
+        )
+
+    def forward(self, x):
+        return self.layer(x)
+
+
+class PositionalEncoding(nn.Module):
+    """
+    Pre-compute position encodings.
+    In forward pass, this adds the position-encodings to the
+    input for as many time steps as necessary.
+    """
+
+    def __init__(self, d_model, dropout, max_len=5000):
+        """
+
+        :param d_model: dimensionality of the model
+        :param dropout:
+        :param max_len: maximum length for which  to pre-compute PE
+        """
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        # Compute the positional encodings once in log space.
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float32).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2, dtype=torch.float32) *
+                             -(math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        """
+        Add PE to x.
+        :param x:
+        :return: x with position encodings added.
+        """
+        x = x + self.pe[:, :x.size(1)]
+        return self.dropout(x)
+
+
+class NoamScheduler:
```

Comment:
Might be worth some explanation, at least the pointer to Eq. 1 in the [Transformer paper](https://arxiv.org/pdf/1706.03762.pdf).

Reply:
We could also consider renaming it to something like `WarmUpCool` or the like.

Reply:
I added a reference to the paper. By now I think many know this one by NoamScheduler I think so let's keep the name for now.